dbms/src/Storages/StorageFile.cpp

```diff
@@ -191,12 +256,23 @@ BlockInputStreams StorageFile::read(
     size_t max_block_size,
     unsigned /*num_streams*/)
 {
-    BlockInputStreamPtr block_input = std::make_shared<StorageFileBlockInputStream>(*this, context, max_block_size);
     const ColumnsDescription & columns_ = getColumns();
     auto column_defaults = columns_.getDefaults();
-    if (column_defaults.empty())
-        return {block_input};
-    return {std::make_shared<AddingDefaultsBlockInputStream>(block_input, column_defaults, context)};
+    if (!path_with_globs)
+    {
+        BlockInputStreamPtr block_input = std::make_shared<StorageFileBlockInputStream>(*this, context, max_block_size, path);
+        if (column_defaults.empty())
+            return {block_input};
+        return {std::make_shared<AddingDefaultsBlockInputStream>(block_input, column_defaults, context)};
+    }
```

Comment:
Looks the same as the loop body below. Can we always use matched_paths, and if we don't have globs, just add the original path there? We won't need `path_with_globs` then.